File: C++/19. Remove Nth Node From End of List.cpp
ListNode* removeNthFromEnd(ListNode* head, int n) 
{
    ListNode *fast = head ; 

    while(n--)
        fast = fast->next ; 
    
    ListNode *prev = head ; 

    while(fast and fast->next)
    {
        fast = fast->next ; 
        prev = prev->next ; 
    }

    if(fast == NULL)
    {
        fast = head ; 
        head = head->next ; 
    }
    else
    {
        fast = prev->next ; 
        prev->next = prev->next->next ; 
    } 
    
    fast->next = NULL ; 
    delete fast ; 

    return head ; 
}

File: Java/19. Remove Nth Node From End of List.java
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode dummyHead = new ListNode(-1, head);

        for (int i = 0; i < n; i++)
            head = head.next;

        ListNode deleteNode = dummyHead;
        while (head != null) {
            head = head.next;
            deleteNode = deleteNode.next;
        }

        deleteNode.next = deleteNode.next.next;

        return dummyHead.next;
    }

    public class ListNode {
        public int val;
        public ListNode next;

        public ListNode() {
        }

        public ListNode(int val) {
            this.val = val;
        }

        public ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }
    }
}